File: backend/app/models/transcode.py
"""Transcode models for external transcoding service."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from sqlalchemy import JSON, Column, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from app.database import Base


class TranscodeNode(Base):
    """Transcode node (worker server) that executes transcoding tasks."""

    __tablename__ = "transcode_nodes"

    id = Column(String(255), primary_key=True)
    name = Column(String(255), nullable=False)
    region = Column(String(50), nullable=False)  # beijing/shanghai/guangzhou
    ip_address = Column(String(50))
    status = Column(String(50))  # online/offline/busy
    max_tasks = Column(Integer, default=4)
    current_tasks = Column(Integer, default=0)
    cpu_usage = Column(Float)
    memory_usage = Column(Float)
    gpu_usage = Column(Float)
    network_latency = Column(Integer)  # latency to SRS Origin in ms
    last_heartbeat = Column(DateTime)
    capabilities = Column(JSON)  # {protocols: [rtmp, webrtc, srt], codecs: [h264, h265], gpu: bool}

    tasks = relationship("TranscodeTask", back_populates="node")


class TranscodeProfile(Base):
    """Transcode configuration profile/template."""

    __tablename__ = "transcode_profiles"

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(255), nullable=False)
    description = Column(Text)
    source_protocol = Column(String(50))  # rtmp/srt/whip
    outputs = Column(JSON)  # [{protocol, resolution, bitrate, fps, codec, audio_codec, audio_bitrate}]
    latency_mode = Column(String(50))  # ultra_low/low/normal
    created_at = Column(DateTime, default=datetime.utcnow)

    tasks = relationship("TranscodeTask", back_populates="profile")


class TranscodeTask(Base):
    """Transcode task instance."""

    __tablename__ = "transcode_tasks"

    id = Column(Integer, primary_key=True, autoincrement=True)
    stream_name = Column(String(255), nullable=False)
    profile_id = Column(Integer, ForeignKey("transcode_profiles.id"), nullable=False)
    node_id = Column(String(255), ForeignKey("transcode_nodes.id"))
    source_protocol = Column(String(50))  # rtmp/srt/whip
    source_url = Column(String(512))
    outputs = Column(JSON)  # [{protocol, bitrate, resolution, url}]
    status = Column(String(50))  # pending/running/stopped/failed
    started_at = Column(DateTime)
    stopped_at = Column(DateTime)
    error_message = Column(Text)
    metrics = Column(JSON)  # {latency, fps, bitrate, packet_loss}

    profile = relationship("TranscodeProfile", back_populates="tasks")
    node = relationship("TranscodeNode", back_populates="tasks")

File: frontend/src/api/transcode.ts
import axios from 'axios';

const API_BASE = '/api/admin/transcode';

export interface TranscodeNode {
  id: string;
  name: string;
  region: string;
  ip_address?: string;
  status: string;
  max_tasks: number;
  current_tasks: number;
  cpu_usage?: number;
  memory_usage?: number;
  gpu_usage?: number;
  network_latency?: number;
  last_heartbeat?: string;
  capabilities?: {
    protocols?: string[];
    codecs?: string[];
    gpu?: boolean;
  };
}

export interface TranscodeProfile {
  id: number;
  name: string;
  description?: string;
  source_protocol: string;
  outputs: Array<{
    protocol: string;
    resolution: string;
    bitrate: string;
    fps: number;
    codec?: string;
    audio_codec?: string;
    audio_bitrate?: string;
  }>;
  latency_mode: string;
  created_at?: string;
}

export interface TranscodeTask {
  id: number;
  stream_name: string;
  profile_id: number;
  node_id?: string;
  source_protocol?: string;
  source_url?: string;
  outputs?: Array<{
    protocol: string;
    url: string;
    resolution?: string;
    bitrate?: string;
    fps?: number;
  }>;
  status: string;
  started_at?: string;
  stopped_at?: string;
  error_message?: string;
  metrics?: {
    latency?: number;
    fps?: number;
    bitrate?: number;
    packet_loss?: number;
  };
}

export const transcodeApi = {
  // Nodes
  listNodes: () => axios.get<{ nodes: TranscodeNode[] }>(`${API_BASE}/nodes`),
  registerNode: (data: Partial<TranscodeNode>) => axios.post(`${API_BASE}/nodes/register`, data),
  deleteNode: (nodeId: string) => axios.delete(`${API_BASE}/nodes/${nodeId}`),

  // Profiles
  listProfiles: () => axios.get<{ profiles: TranscodeProfile[] }>(`${API_BASE}/profiles`),
  getProfile: (id: number) => axios.get<TranscodeProfile>(`${API_BASE}/profiles/${id}`),
  createProfile: (data: Partial<TranscodeProfile>) => axios.post(`${API_BASE}/profiles`, data),
  updateProfile: (id: number, data: Partial<TranscodeProfile>) =>
    axios.put(`${API_BASE}/profiles/${id}`, data),
  deleteProfile: (id: number) => axios.delete(`${API_BASE}/profiles/${id}`),

  // Tasks
  listTasks: (params?: { region?: string; status?: string }) =>
    axios.get<{ tasks: TranscodeTask[] }>(`${API_BASE}/tasks`, { params }),
  getTask: (id: number) => axios.get<TranscodeTask>(`${API_BASE}/tasks/${id}`),
  createTask: (data: { stream_name: string; profile_id: number; region?: string; auto_start?: boolean }) =>
    axios.post(`${API_BASE}/tasks`, data),
  startTask: (id: number) => axios.post(`${API_BASE}/tasks/${id}/start`),
  stopTask: (id: number) => axios.post(`${API_BASE}/tasks/${id}/stop`),
  deleteTask: (id: number) => axios.delete(`${API_BASE}/tasks/${id}`),

  // Regions
  listRegions: () => axios.get<{ regions: Array<{
    name: string;
    display_name: string;
    nodes: number;
    online_nodes: number;
    total_capacity: number;
    used_capacity: number;
    avg_latency: number;
  }> }>(`${API_BASE}/regions`),
};
